executor/union_scan.go

```diff
@@ -269,6 +262,49 @@ func (us *UnionScanExec) getSnapshotRow(ctx context.Context) ([]types.Datum, err
 	return us.snapshotRows[0], nil
 }
 
+func (us *UnionScanExec) isSnapshotRowConflictWithMemBufRow(row chunk.Row) (bool, error) {
+	// Check conflict by handle.
+	snapshotHandle, err := us.belowHandleCols.BuildHandle(row)
+	if err != nil {
+		return false, err
+	}
+	var checkKey kv.Key
+	if us.physTblIDIdx >= 0 {
+		tblID := row.GetInt64(us.physTblIDIdx)
+		checkKey = tablecodec.EncodeRowKeyWithHandle(tblID, snapshotHandle)
+	} else {
+		checkKey = tablecodec.EncodeRecordKey(us.table.RecordPrefix(), snapshotHandle)
+	}
+	if _, err := us.memBufSnap.Get(context.TODO(), checkKey); err == nil {
+		// If src handle appears in added rows, it means there is conflict and the transaction will fail to
+		// commit, but for simplicity, we don't handle it here.
+		return true, nil
+	}
+
+	// Check conflict by unique index.
+	stmtCtx := us.ctx.GetSessionVars().StmtCtx
+	for _, idxCols := range us.uniqueIndexCols {
+		idxVals := idxCols.FetchIndexValues(row)
+		// Pass handle = 0 to GenIndexKey, due to we only care about distinct key.
+		iter := idxCols.IdxInfo.GenIndexKVIter(stmtCtx, idxVals, kv.IntHandle(0), nil)
+		for iter.Valid() {
+			uniqueKey, _, distinct, err := iter.Next(nil)
```

Comment:
We also need to take the physical table id into account when building a unqiue key. Considering the following case:
```sql
/* t1 */ drop table if exists t;
/* t1 */ create table t (c1 int, c2 int, primary key (c1) nonclustered) partition by hash(c1) partitions 4;
/* t1 */ insert into t values (1, 1);
/* t1 */ begin;
/* t1 */ update t set c1=2 where c1=1;
/* t2 */ begin;
/* t2 */ insert into t values (1, 2);
/* t1 */ commit;
/* t2 */ select * from t use index(primary) order by c1, c2;  -- [(1, 1), (1, 2)]
```